File: program/program/implement-ternary-search/implement_ternary_search.go
package main

import "fmt"

func main() {
	list := []int{1, 2, 4, 5, 9, 10, 12, 13, 15, 16}
	value := 12
	index := ternary_search(list, value)

	if index != -1 {
		fmt.Printf("Element %d found at index %d", value, index)
	} else {
		fmt.Printf("Element %d not found !", value)
	}
}

func ternary_search(array []int, value int) int {
	var l = 0
	var r = len(array) - 1

	for l <= r {
		ps := (r - l) / 3
		m1 := l + ps
		m2 := r - ps

		if array[m1] == value {
			return m1
		}

		if array[m2] == value {
			return m2
		}

		if value < array[m1] {
			r = m1 - 1
		} else if value > array[m2] {
			l = m2 + 1
		} else {
			l = m1 + 1
			r = m2 - 1
		}
	}

	return -1

}

File: program/program/implement-ternary-search/implement_ternary_search.r
ternary_search <- function(A, value) 
  {
  l <- 1
  r <- length(A)
  
  while (l <= r) {
    partition_size <- (r - l) / 3
    mid1 <- floor(l + partition_size)
    mid2 <- ceiling(r - partition_size)
    
    if (A[mid1] == value) 
      {
      return(mid1 - 1)
    }
    
    if (A[mid2] == value) 
      {
      return(mid2 - 1)
    }
    
    if (value < A[mid1]) 
      {
      r <- mid1 - 1
    } else if (value > A[mid2]) 
      {
      l <- mid2 + 1
    } else 
      {
      l <- mid1 + 1
      r <- mid2 - 1
    }
  }
  
  return(-1)
}

sorted_array <- c(1, 2, 3, 4, 5)
search_value <- 4
result <- ternary_search(sorted_array, search_value)

if (result == -1) 
  {
  cat("ERROR! The given element does not exist in the current array :(\n")
} else 
  {
  cat("Element found! It is located at index", result, "\n")
}

# Cautiously Crafted By Redzwinger #
